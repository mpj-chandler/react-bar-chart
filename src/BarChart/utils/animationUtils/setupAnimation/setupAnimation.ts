import React from 'react';
import onStart from '../onStart/onStart';
import loop from '../loop/loop';
import onFrame from '../onFrame/onFrame';

export function setupAnimation(setTime: React.Dispatch<React.SetStateAction<number>>, duration: number, delay: number) {
    let animationFrame: number;
    const start: number = Date.now();

    animationFrame = loop(onFrame(start, duration, setTime));

    const timerStop = onStart(setTime, start, animationFrame, duration);
    const timerDelay = setTimeout(() => onStart(setTime, start, animationFrame, duration), delay);

    return () => {
        clearTimeout(timerStop);
        clearTimeout(timerDelay);
        cancelAnimationFrame(animationFrame);
    };
}


