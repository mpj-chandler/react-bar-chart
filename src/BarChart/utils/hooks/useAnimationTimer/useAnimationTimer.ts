import { useEffect, useState } from 'react';
import { setupAnimation } from '../../animationUtils/setupAnimation/setupAnimation';

export function useAnimationTimer(duration: number, delay: number) {

    const [elapsed, setTime] = useState(0);

    useEffect(() => {
        setupAnimation(setTime, duration, delay);
    },
    [duration, delay],
    );

    return  elapsed;
}
