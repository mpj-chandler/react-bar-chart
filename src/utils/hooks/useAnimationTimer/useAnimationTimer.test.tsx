import React from 'react';
import TestRenderer from 'react-test-renderer';
import { useAnimationTimer } from './useAnimationTimer';

import { setupAnimation } from '../../animationUtils/setupAnimation/setupAnimation';

jest.mock('../../animationUtils/setupAnimation/setupAnimation');

const MockUseAnimationExecutor: (props: {useAnimationTimer: (duration: number, delay: number) => number, duration: number, delay: number}) => JSX.Element = (props) => {
    const elapsed = props.useAnimationTimer(props.duration, props.delay);
    return <div>{elapsed}</div>;
}

const mockComponentFactory = (duration: number, delay: number) => {
    return TestRenderer.create(<MockUseAnimationExecutor useAnimationTimer={useAnimationTimer} duration={duration} delay={delay}/>);
};

describe('the useAnimationTimer hook', () => {
    afterEach(jest.resetAllMocks);

    [0, 500, 1000].forEach((duration: number) => {
        it(`it renders consistently with duration: ${duration}`, () => {
            const mockComponent = mockComponentFactory(duration, 0);
            expect(mockComponent).toMatchSnapshot();
        });
    });

    [0, 500, 1000].forEach((delay: number) => {
        it(`it renders consistently with delay: ${delay}`, () => {
            const mockComponent = mockComponentFactory(1000, delay);
            expect(mockComponent).toMatchSnapshot();
        });
    });
    
    it('it responds to changes in the duration prop', () => {
        const mockComponent = mockComponentFactory(0, 0);

        expect(setupAnimation).toHaveBeenCalledTimes(1);
        expect(setupAnimation).toHaveBeenCalledWith(expect.any(Function), 1000, 1000);

        mockComponent.update(<MockUseAnimationExecutor useAnimationTimer={useAnimationTimer} duration={1000} delay={0}/>);

        expect(setupAnimation).toHaveBeenCalledTimes(2);
        expect(setupAnimation).toHaveBeenCalledWith(expect.any(Function), 0, 0);
    })

});