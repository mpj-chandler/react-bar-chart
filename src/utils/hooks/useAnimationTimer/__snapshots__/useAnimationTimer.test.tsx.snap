// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the useAnimationTimer hook it renders consistently with delay: 0 1`] = `
<div>
  0
</div>
`;

exports[`the useAnimationTimer hook it renders consistently with delay: 500 1`] = `
<div>
  0
</div>
`;

exports[`the useAnimationTimer hook it renders consistently with delay: 1000 1`] = `
<div>
  0
</div>
`;

exports[`the useAnimationTimer hook it renders consistently with duration: 0 1`] = `
<div>
  0
</div>
`;

exports[`the useAnimationTimer hook it renders consistently with duration: 500 1`] = `
<div>
  0
</div>
`;

exports[`the useAnimationTimer hook it renders consistently with duration: 1000 1`] = `
<div>
  0
</div>
`;
